packages/blade/src/components/ActionList/ActionList.tsx

```diff
@@ -1,119 +1,149 @@
-  const actionListOptions: {
-    title: string;
-    value: string;
-  }[] = [];
+      const clonedChild = React.cloneElement(child, {
+        // @ts-expect-error: TS doesn't understand the child's props
+        _index: currentIndex,
+      });
+      return clonedChild;
+    }
 
-  console.count('ActionList');
+    return child;
+  };
+
+  const childCompIdArray = React.Children.toArray(children).map((child) => getComponentId(child));
+  const lastActionListSectionIndex = childCompIdArray.lastIndexOf(componentIds.ActionListSection);
+
+  const isActionListItemPresentAfterSection = childCompIdArray
+    .slice(lastActionListSectionIndex)
+    .includes(componentIds.ActionListItem);
 
   // Looping through ActionListItems to add index to them and get an options array for moving focus between items
-  const childrenWithId = React.Children.map(children, (child) => {
+  const childrenWithId = React.Children.map(children, (child, index) => {
     if (React.isValidElement(child)) {
-      // @TODO: handle the scenario where ActionListItem is inside ActionListMenu
-      if (child.type === ActionListItem) {
-        actionListOptions.push({
-          title: child.props.title,
-          value: child.props.value,
-        });
-        const currentIndex = actionListOptions.length - 1;
-        const clonedChild = React.cloneElement(child, {
+      if (isValidAllowedChildren(child, componentIds.ActionListSection)) {
+        return React.cloneElement(child, {
           // @ts-expect-error: TS doesn't understand the child's props
-          _index: currentIndex,
+          children: React.Children.map(child.props.children, (childInSection) => {
+            if (isValidAllowedChildren(childInSection, componentIds.ActionListItem)) {
+              return getActionListItemChildWithId(childInSection);
+            }
+
+            return childInSection;
+          }),
+          hideDivider: index === lastActionListSectionIndex && !isActionListItemPresentAfterSection,
         });
-        return clonedChild;
       }
-    }
 
+      if (isValidAllowedChildren(child, componentIds.ActionListItem)) {
+        return getActionListItemChildWithId(child);
+      }
+    }
     return child;
   });
 
   React.useEffect(() => {
     setOptions(actionListOptions);
     // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [setOptions, children]);
+  }, [optionsRecalculateToggle]);
+
+  React.useEffect(() => {
+    setSelectedIndices(defaultSelectedIndices);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
 
   return (
-    <Box
+    <StyledActionList
       // eslint-disable-next-line @typescript-eslint/no-explicit-any
       ref={actionListRef as any}
       {...makeAccessible({
-        role: 'listbox',
+        role: isReactNative() ? 'menu' : 'listbox',
         multiSelectable: selectionType === 'multiple',
         labelledBy: `${dropdownBaseId}-label`,
       })}
       id={`${dropdownBaseId}-listbox`}
```

Comment:
I moved the a11y things into separate `getA11yRoles.ts` file for now so that we can control roles from outside of all this code.

By generic if you mean generic for menus, then it is right now generic for menus. 

But by generic if you mean, usable as a plain list somewhere then the accessibility completely changes. 90% of the code in ActionList won't be used in that scenario. We can go several ways there
- Have a separate Navigation component (this would be ideal for bundle size, and managing code accessibility and behaviour for us)
- Add if else checks to not run some of the code (click events, selection calls, keydown events, the runtime logic to add index to items, etc)

Lets discuss this after standup today.